File: go-scanner/scanner/scanner.go
package scanner

import (
	"fmt"
	"strconv"
	"unicode"
	"unicode/utf8"

	"github.com/jkbcz/yacub/token"
)

type Scanner struct {
	source []byte

	current   int
	start     int
	line      int
	lineStart int
}

func NewScanner(source []byte) *Scanner {

	return &Scanner{
		source:    source,
		current:   0,
		start:     0,
		line:      1,
		lineStart: 0,
	}
}

func (s *Scanner) ScanAll() ([]token.Token, error) {
	var tokens []token.Token
	for {
		t, err := s.Scan()
		if err != nil {
			return tokens, err
		}
		tokens = append(tokens, t)
		if t.Type == token.EOF {
			return tokens, nil
		}
	}
}

func (s *Scanner) Scan() (token.Token, error) {
	s.start = s.current

	if s.isAtEnd() {
		return s.token(token.EOF, nil), nil
	}
	ch := s.advance()
	switch ch {
	case '(':
		return s.token(token.LEFT_PAREN, nil), nil
	case ')':
		return s.token(token.RIGHT_PAREN, nil), nil
	case '-':
		return s.token(token.MINUS, nil), nil
	case '+':
		return s.token(token.PLUS, nil), nil
	case '*':
		return s.token(token.STAR, nil), nil
	case '/':
		return s.token(token.SLASH, nil), nil
	case '!':
		return s.token(s.match2('=', token.BANG_EQUAL, token.BANG), nil), nil
	case '=':
		return s.token(s.match2('=', token.EQUAL_EQUAL, token.EQUAL), nil), nil
	case '<':
		return s.token(s.match2('=', token.LESS_EQUAL, token.LESS), nil), nil
	case '>':
		return s.token(s.match2('=', token.GREATER_EQUAL, token.GREATER), nil), nil
	case ' ':
	case '\r':
	case '\t':
		break
	case '\n':
		s.line++
		s.lineStart = s.current
	default:
		if isDigit(ch) {
			return s.scanNumber()
		}
		if isAlpha(ch) {
			return s.scanIdentifier()
		}
		return token.Token{}, s.errorf("invalid character: '%c'", ch)
	}

	return s.Scan()
}

func (s *Scanner) scanNumber() (token.Token, error) {
	for isDigit(s.peek()) {
		s.advance()
	}

	value, err := strconv.Atoi(string(s.source[s.start:s.current]))
	return s.token(token.NUMBER, value), err
}

func (s *Scanner) scanIdentifier() (token.Token, error) {
	for isAlphaNumeric(s.peek()) {
		s.advance()
	}
	return s.token(token.IDENTIFIER, nil), nil
}

func (s *Scanner) advance() rune {
	r, size := utf8.DecodeRune(s.source[s.current:])
	s.current += size
	return r
}

func (s *Scanner) peek() rune {
	r, _ := utf8.DecodeRune(s.source[s.current:])
	return r
}

func (s *Scanner) match2(expected rune, t1, t2 token.Type) token.Type {
	if s.isAtEnd() {
		return t2
	}
	if s.peek() != expected {
		return t2
	}
	s.advance()
	return t1
}

func (s *Scanner) isAtEnd() bool {
	return s.current >= len(s.source)
}

func (s *Scanner) token(typ token.Type, lit any) token.Token {
	lexeme := string(s.source[s.start:s.current])
	return token.Token{
		Type:    typ,
		Lexeme:  lexeme,
		Literal: lit,
		Line:    s.line,
		Column:  s.column(),
	}
}

func (s *Scanner) error(msg string) Error {
	return Error{
		Line:   s.line,
		Column: s.column(),
		Msg:    msg,
	}
}

func (s *Scanner) column() int {
	return s.start - s.lineStart + 1
}

func (s *Scanner) errorf(msgFormat string, args ...any) Error {
	return s.error(fmt.Sprintf(msgFormat, args...))
}

func isDigit(a rune) bool {
	return a >= '0' && a <= '9'
}

func isAlpha(a rune) bool {
	return unicode.IsLetter(a)
}

func isAlphaNumeric(a rune) bool {
	return isAlpha(a) || isDigit(a)
}

File: syntax-highlighter/src/scanner.rs
use std::fs;

use crate::token::{Token, TokenType};

pub struct Scanner {
    source: String,
    current: usize,
    start: usize,
    line: usize,
    line_start: usize,
}

impl Scanner {
    pub fn new(filename: String) -> Result<Self, std::io::Error> {
        let source = fs::read_to_string(filename)?;

        Ok(Scanner {
            source,
            current: 0,
            start: 0,
            line: 1,
            line_start: 0,
        })
    }

    pub fn source(&self) -> &str {
        &self.source
    }

    pub fn scan_all(&mut self) -> Result<Vec<Token>, String> {
        let mut tokens = Vec::new();

        loop {
            let token = self.scan()?;
            let is_eof = token.token_type == TokenType::Eof;
            tokens.push(token);

            if is_eof {
                break;
            }
        }

        Ok(tokens)
    }

    fn scan(&mut self) -> Result<Token, String> {
        self.start = self.current;

        if self.is_at_end() {
            return Ok(self.token(TokenType::Eof, None));
        }

        let ch = self.advance();
        match ch {
            '(' => Ok(self.token(TokenType::LeftParen, None)),
            ')' => Ok(self.token(TokenType::RightParen, None)),
            '-' => Ok(self.token(TokenType::Minus, None)),
            '+' => Ok(self.token(TokenType::Plus, None)),
            '*' => Ok(self.token(TokenType::Star, None)),
            '/' => Ok(self.token(TokenType::Slash, None)),
            '!' => {
                let typ = self.match2('=', TokenType::BangEqual, TokenType::Bang);
                Ok(self.token(typ, None))
            }
            '=' => {
                let typ = self.match2('=', TokenType::EqualEqual, TokenType::Equal);
                Ok(self.token(typ, None))
            }
            '<' => {
                let typ = self.match2('=', TokenType::LessEqual, TokenType::Less);
                Ok(self.token(typ, None))
            }
            '>' => {
                let typ = self.match2('=', TokenType::GreaterEqual, TokenType::Greater);
                Ok(self.token(typ, None))
            }
            ' ' | '\r' | '\t' => self.scan(), // Skip whitespace, recursively scan next token
            '\n' => {
                self.line += 1;
                self.line_start = self.current;
                self.scan() // Recursively scan next token
            }
            '0'..='9' => self.scan_number(),
            ch if ch.is_alphabetic() => self.scan_identifier(),
            _ => Err(self.error(format!("Invalid character: '{}'", ch))),
        }
    }

    fn advance(&mut self) -> char {
        let ch = self.source[self.current..].chars().next().unwrap();
        self.current += ch.len_utf8();
        ch
    }

    fn peek(&self) -> Option<char> {
        self.source[self.current..].chars().next()
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    fn scan_number(&mut self) -> Result<Token, String> {
        while let Some(ch) = self.peek() {
            if ch.is_ascii_digit() {
                self.advance();
            } else {
                break;
            }
        }

        let lexeme = &self.source[self.start..self.current];

        let value = lexeme
            .parse::<i32>()
            .map_err(|e| self.error(format!("Invalid number: {}", e)))?;

        Ok(self.token(TokenType::Number, Some(value)))
    }

    fn scan_identifier(&mut self) -> Result<Token, String> {
        while let Some(ch) = self.peek() {
            if ch.is_alphanumeric() || ch == '_' {
                self.advance();
            } else {
                break;
            }
        }

        Ok(self.token(TokenType::Identifier, None))
    }

    fn match2(&mut self, expected: char, if_match: TokenType, if_not: TokenType) -> TokenType {
        if self.is_at_end() {
            return if_not;
        }

        if self.peek() != Some(expected) {
            return if_not;
        }

        self.advance();
        if_match
    }

    fn token(&self, typ: TokenType, literal: Option<i32>) -> Token {
        let lexeme = self.source[self.start..self.current].to_string();
        Token {
            token_type: typ,
            lexeme,
            literal,
            line: self.line,
            column: self.column(),
        }
    }

    fn column(&self) -> usize {
        self.start - self.line_start + 1
    }

    fn error(&self, msg: String) -> String {
        format!("{}:{}: {}", self.line, self.column(), msg)
    }
}
